core/src/main/java/org/elasticsearch/common/io/stream/NamedWriteableRegistry.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.io.stream;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Registry for {@link NamedWriteable} objects. Allows to register and retrieve prototype instances of writeable objects
+ * given their name.
+ */
+public class NamedWriteableRegistry {
+
+    private Map<String, NamedWriteable> registry = new HashMap<>();
+
+    /**
+     * Registers a {@link NamedWriteable} prototype
+     */
+    public synchronized void registerPrototype(NamedWriteable<?> namedWriteable) {
+        if (registry.containsKey(namedWriteable.getWriteableName())) {
+            throw new IllegalArgumentException("named writeable of type [" + namedWriteable.getClass().getName() + "] with name [" + namedWriteable.getWriteableName() + "] " +
+                    "is already registered by type [" + registry.get(namedWriteable.getWriteableName()).getClass().getName() + "]");
+        }
+        registry.put(namedWriteable.getWriteableName(), namedWriteable);
+    }
+
+    /**
+     * Returns a prototype of the {@link NamedWriteable} object identified by the name provided as argument
+     */
+    public <C> NamedWriteable<C> getPrototype(String name) {
+        @SuppressWarnings("unchecked")
+        NamedWriteable<C> namedWriteable = (NamedWriteable<C>)registry.get(name);
```

Comment:
I think we should avoid implicit casts and instead expect consumers of the API to cast themselves. Can we just make this method return a NamedWriteable<?>?